### PR TITLE
Targeted detailed recording

### PR DESCRIPTION
**Pull request by adelowo**
Fixes https://github.com/TykTechnologies/product/issues/168

- [x] Add flag on API Definition
- [x] Flag on key level

### CODE REVIEW

File: apidef/api_definitions.go [hunk 1]
```diff
@@ -469,15 +469,16 @@ type APIDefinition struct {
...
+	Domain                  string                 `bson:"domain" json:"domain"`
+	Certificates            []string               `bson:"certificates" json:"certificates"`
+	DoNotTrack              bool                   `bson:"do_not_track" json:"do_not_track"`
+	Tags                    []string               `bson:"tags" json:"tags"`
+	EnableContextVars       bool                   `bson:"enable_context_vars" json:"enable_context_vars"`
+	ConfigData              map[string]interface{} `bson:"config_data" json:"config_data"`
+	TagHeaders              []string               `bson:"tag_headers" json:"tag_headers"`
+	GlobalRateLimit         GlobalRateLimit        `bson:"global_rate_limit" json:"global_rate_limit"`
+	StripAuthData           bool                   `bson:"strip_auth_data" json:"strip_auth_data"`
+	EnableDetailedAnalytics bool                   `bson:"enable_detailed_recording" json:"enable_detailed_recording"`
```
Comment (gernest) [inline apidef/api_definitions.go]: `EnableDetailedAnalytics` doesn't reflect `enable_detailed_recording` I think we need to be consistent here and name it `EnableDetailedRecording`

Comment (adelowo) [reply apidef/api_definitions.go]: Fixed.

**Comment by buger** [comment]
So, we already had `EnableDetailedRecording` on session level (just not used), correct?

**Comment by adelowo** [comment]
Yes, we only use it when it is set and organisation tracking/logging is enabled

**Comment by buger** [comment]
So to clarify one more time, we have 3 ways to enable detailed logging now:
1) API level
2) Session level
3) Org level (org keys)

Correct?

**Comment by adelowo** [comment]
4. There's also the global config option

File: apidef/api_definitions.go (hunk 1 above)
Comment (letzya) [inline apidef/api_definitions.go]: so the field is `enable_detailed_recording` and the variable is EnableDetailed`Analytics`?

Comment (adelowo) [reply apidef/api_definitions.go]: You seem to have reviewed an old commit or something. It was updated before this PR was merged. https://github.com/TykTechnologies/tyk/commit/5e74e19c6759f25d6edc2544b1f88f18634d9a7a

**Comment by letzya** [comment]
@adelowo would be good to add some info in the description of the pr explaining a bit the logic and how these config being applied.
This later can be use in the docs or for early adopters of users of the gateway

File: apidef/api_definitions.go (hunk 1 above)
Comment (letzya) [reply apidef/api_definitions.go]: Thanks! Got this link from the release channel

**Comment by letzya** [comment]
@adelowo This is the logic we need to follow: 
image

**Comment by lghiur** [comment]
@adelowo please provide more info about this in the docs, as per @adelowo comment above

**Comment by adelowo** [comment]
Also, docs are at https://github.com/TykTechnologies/tyk-docs/pull/1235

**Comment by adelowo** [comment]
targeted_detailed_recording-2

**Pull request merged**